### DISCUSSION TITLE
Include battery (storage) control mode as part of the optimisation data frame

### DISCUSSION THREAD

**Discussion by fredriklj** [davidusb-geek/emhass#360] [Ideas]
Many inverters connected to a storage (battery) can be controlled according to Sunspec (true for at least fronius and solaredge). There is a quite good explanation of how it works here:

https://loxwiki.atlassian.net/wiki/spaces/LOXEN/pages/1316061809/Fronius+Hybrid+with+Modbus+TCP#[hardBreak]2.-Control-the-battery-charge-rate-and-discharge-rate.

This can, somewhat simplified, be translated into six useful modes of operation for your battery storage. These are:

1. Self sufficient mode: the inverter will (based on readings from a meter) balance the charging and discharging of the battery as to minimise transmission to/from the grid based on consumption and PV production. This could be regarded as a default mode,

2. Prevent discharging: the inverter will charge the storage until full and export surplus from PV, but not draw any current from the storage. This is useful during times when the energy prices are low (it is not worth the wear of the battery to use it). Effectively this is a type of idle state, where only surplus energy is used for charging.

3. Prevent charging: the inverter will draw current from the storage to minimise transmission from the grid, but will export any/all surplus generated from PV. This mode is useful when energy prices are high to maximise export of surplus from PV.

4. Charging: the inverter will charge the battery with the specified current (up to what the inverter and battery is capable of), irrespectively of consumption or PV production. This mode is useful when energy prices are low, to charge the battery to a certain SoC for later use of the energy.

5. Discharging: the inverter will discharge the battery with the specified current (up to what the inverter and battery is capable of), irrespectively of consumption or PV production. This mode is useful when energy prices are very high, to maximise export of energy to the grid.

6. Idle: the inverter will neither charge nor discharge the battery, irrespectively of consumption or PV production. This mode is mostly useful for long periods of low energy prices, to minimise wear and depletion of the storage.

I have been trying to sort out from MPC optimisation when to switch between the modes. It is not too hard, but there is a slight room for errors when predictions are off and it feels like much of this is already available from the optimisation model.

It should, I think, be fairly easy to provide an additional column in the data frame to identify mode 4 (charging from the grid to the battery), and mode 5 (when to discharge from the battery to the grid). It would also be useful to know if charging/discharging should be 100% of what the system is capable of or less. This could be expressed as percentage or power of maximal capacity.

Mode 3 (prevent charging) could be selected when the current export price is higher than the import price later. Could be more optimal to export PV production when the sun shines and then charge the battery from the grid later.

When the import energy price falls below "weight_battery_discharge", then mode 2 can be selected. No reason to discharge the battery when it is cheaper to import from the grid.

I think there is no use for the idle mode in the optimisation, so it can be disregarded. 

It would have been extremely nice to get these modes out from the optimisation directly, instead of trying to reverse the battery power column into this.

Any thoughts?

**Comment by scruysberghs** [comment]
I use a template to calculate a mode from the emhass sensors :
template:
  - sensor:
      - name: "EMHASS batcontrol mode"
        unique_id: emhass_batcontrol_mode
        icon: mdi:battery-heart-outline
        state: >
          {% set p_hybrid = states('sensor.mpc_p_hybrid_inverter')| float(0) %}       
          {% set p_batt = states('sensor.mpc_p_batt_forecast') | float(0) %}
          {% set p_load = states('sensor.mpc_p_load_forecast') | float(0) %}     
          {% set p_grid = states('sensor.mpc_p_grid_forecast') | float(0) %} 
          {% set surplus = states('sensor.surplus_power') | float(0) %}
          {% set max_captar = states('input_number.max_power_from_grid') | float(0) %} 
          {% set captar15 = states('sensor.sec_pac_extrapoleer_captar_kwartier') | float(0) %}
          {% if p_grid == 0 or (p_batt <= 0 and surplus > 20 and p_grid == 0) %} 
            Follow Grid
          {% elif p_grid >= max_captar and p_load < max_captar  %}
            Charge Maxcaptar
          {% elif p_grid >= max_captar and p_load > max_captar %}
            Peak Shaving   
          {% elif p_batt == 0 %}
            Do Nothing
          {% elif p_batt < 0 %}
            Forcible Charge
          {% elif p_batt > 0 %}
            Forcible Discharge
          {% else %}
            Do Nothing
          {% endif %}
It then use the resulting mode in an automation/script to make the inverter do the right thing, set limits, set forcible charge/discharge,. If the inverter has it's own powermeter you can use it's "maximise self consumtion" mode to "follow grid", if there is no power meter you can do this by settings charge en discharge power yourself from what home assistant knows is coming from solar, grid and battery

**Reply by fredriklj** [reply]
> I use a template to calculate a mode from the emhass sensors : template:

Right! I have done something similar, looking like this:

      {% if (states('sensor.emhass_p_batt_forecast') | int(0)) > 0 and (states('sensor.emhass_p_grid_forecast') | int(0)) < 0 %}
        Discharging
      {% elif (states('sensor.emhass_p_batt_forecast') | int(0)) < 0 and (states('sensor.emhass_p_grid_forecast') | int(0)) > 0 %}
        Charging
      {% elif (states('sensor.spotprice_now') | float(0)) < (states('input_number.floor_optimisation') | float(0)) %}
        Prevent discharging
      {% else %}
        Self-sufficient
      {% endif %}


(somewhat edited for readability)

It works pretty well, but I think there are some border cases where it can end up wrong. I compensate for that by throttling the charging/discharging current according to the optimisation made. I think (although have not seen) that the battery can be set to charge or discharge with a small current if the p_batt and p_grid don't quite add up with solar production.

**Comment by anselmhuewe** [comment]
I support these modes in www.shyft-power.com and derive them from the optimization model. "Idle" is pretty useful if you want to preserve remaining power in the battery for later hours where electricity is more expansive. For example, if the optimizer models the SOC to remain unchanged, I put the battery on idle mode to implement the modeled optimal path.

**Reply by fredriklj** [reply]
> "Idle" is pretty useful if you want to preserve remaining power in the battery for later hours where electricity is more expansive.

I actually never set the battery to "Idle", in the case you mention I set it to prevent discharging. In that case any surplus PV production will be used to charge the battery (until full of course). Setting it to "Idle" would have had the effect that surplus PV power would have been exported regardless of SoC.

**Reply by anselmhuewe** [reply]
yes, you are right, sorry for my sloppy wording. I use "maximize self consumption" and set the discharge rate to zero if I want the battery not to discharge. Or I set "charge from solar power and grid" and compute the charge rate to match the result of the optimizer if I want it to charge from the grid.
